### PR TITLE
set the default to use set_to_none for clearing gradients in BF16 optimizer.

### PR DESCRIPTION
**Pull request by inkcherry**
as discussed in #5175,  set the default to use set_to_none for clearing gradients in BF16 optimizer.
Additionally, for the case of zero clearing, use foreach_zero.
Verified correctness with mega-ds llama 7B training.

FYI @loadams

### CODE REVIEW

File: deepspeed/runtime/bf16_optimizer.py [hunk 1]
```diff
@@ -441,11 +441,20 @@ def clear_hp_grads(self):
             self.fp32_groups_has_gradients[i] = [False] * len(group)
 
     def clear_lp_grads(self):
+
+        # using zero_() fixed memory address for graph replay
+        set_to_none = set_to_none = False if self.graph_harvesting else True
```
Comment (tjruwase) [inline deepspeed/runtime/bf16_optimizer.py:446]:
```suggestion
        set_to_none = False if self.graph_harvesting else True
```

Comment (inkcherry) [reply deepspeed/runtime/bf16_optimizer.py]: fixed and reverified

File: deepspeed/runtime/bf16_optimizer.py [hunk 2]
```diff
@@ -341,7 +341,7 @@ def _update_hp_grad(self, lp, group_idx, param_idx, clear_lp_grads):
 
         # clear gradients
         if clear_lp_grads:
-            lp.grad._zero()
+            lp.grad.zero_()
```
Comment (tjruwase) [inline deepspeed/runtime/bf16_optimizer.py:344]: @inkcherry, will this break on older pytorch versions?

Comment (inkcherry) [reply deepspeed/runtime/bf16_optimizer.py:344]: It shouldn't, referring to the definition of ```zero_grad``` in version 1.10 at https://pytorch.org/docs/1.10/_modules/torch/optim/optimizer.html#Optimizer

**Review by tjruwase — APPROVED** [review]
